Freshness section slight reorg (#91)
ietf-rats-wg/architecture: draft-ietf-rats-architecture.md

Attester:

> : An entity whose attributes must be appraised in order to determine whether the entity is considered trustworthy, such as when deciding whether the entity is authorized to perform some operation

Reviewer: I believe "Evidence" is slightly better than "attributes" (but I can live with it either way) since only the attributes that appear in Evidence are actually appraised.

On "whether" vs "the extent to which", I believe the latter is more correct since the former implies it's always exactly boolean, and the latter is more liberal in terms of how may different possibilities there are.

Reviewer 2: +1, I agree with Dave

Author: This change has been reverted.

Endorser:

: An entity that creates Endorsements that can be used to help to appraise the trustworthiness of Attesters

Evidence:

: A set of information about an Attester that is to be appraised by a Verifier

Relying Party:

> : An entity that depends on the validity of information about another entity, typically for purposes of authorization.  Compare /relying party/ in {{RFC4949}}

Reviewer: FYI I am ok either way, but I think "another entity" was previously changed to "an Attester" so not sure why reverting is being proposed in this PR but again I can live with it either way.

Author: This change has been reverted.

: An entity, that depends on the validity of information about an Attester, for purposes of reliably applying application specific actions.  Compare /relying party/ in {{RFC4949}}
: An entity that depends on the validity of information about another entity, typically for purposes of authorization.  Compare /relying party/ in {{RFC4949}}

Relying Party Owner:

: An entity (typically an administrator), that is authorized to configure Appraisal Policy for Attestation Results in a Relying Party
: An entity, such as an administrator, that is authorized to configure Appraisal Policy for Attestation Results in a Relying Party.

Verifier:

: An entity that appraises the validity of Evidence about an Attester

>  and produces Attestation Results to be used by a Relying Party

Reviewer: Why is this line removed?

Author: This change has been reverted.

These components perform a series of measurements, and express this with Evidence as to the
hardware and firmware/software that is running.

~~~
    FIXME from Henk: Measurements at early stages of
    Layered Attestation are NOT evidence yet.
    This text does not cover that yet

> ~~~

Reviewer: We should not be adding FIXME comments into the doc at this point

Author: This change has been reverted.

~~~~
{:twotypes-env #twotypes-env title="Two Types of Environments"}

## Layered Attestation Procedures {#layered-attestation}

By definition, the Attester role takes on the duty to create Evidence.
The fact that an Attester role is composed of several types of environments that
can be nested or staged adds complexity to the architectural layout of how an

> Attester -- in itself -- is composed and therefore has to conduct the Claims collection

Reviewer: I find the wording of "--in itself-- is composed" to be harder to read than the previous text

Author: This change has been reverted.

in order to create believable Attestation Evidence.
The following example is intended to illustrate this composition:

> A very common example is elaborated on to illustrate Layered Attestation.

Author: This has been reverted.

{:layered: artwork-align="center"}
~~~~ LAYERED
{::include layered-attester.txt}
~~~~
{:layered #layered title="Layered Attester"}

The very first Attesting Environment has to ensure the integrity of
the (U)EFI / BIOS / Firmware that initially boots up a composite device (e.g.,
a cell phone).

~~~
    Henk: we are looking for a better term than UEFI/BIOS/Firmware

> ~~~

Reviewer: This seems to revert a bunch of changes we agreed to.   Maybe it needs a rebase or something?

Author: Yes.  All of these changes have now been reverted.

# Freshness {#freshness}

> It is important to prevent old Evidence or old Attestation Results that are no longer correct to be assessed as valid. Replay attacks are an important reason to do that. More importantly, the validity of Conceptual Messages, such as Evidence or Attestation Results, are the basis for trustworthy remote attestation procedures, in general.

Reviewer: Please break lines into shorter source lines, like in the old text, to make diffs easier to review.

Author: Done in b0488b2 for the whole Freshness section.

> It is important to prevent old Evidence or old Attestation Results that are no longer correct to be assessed as valid. Replay attacks are an important reason to do that. More importantly, the validity of Conceptual Messages, such as Evidence or Attestation Results, are the basis for trustworthy remote attestation procedures, in general.

Author: This text doesn't exist anymore.

Some mechanism of ensuring that Evidence and Attestation
Results are actually fresh are required. In essence, there has to be some degree of assurance
that they still reflect the latest operational state of the Attester and that, for example, Attestation
Results were generated using the latest Appraisal Policy for Evidence.
There is, however, always a race condition possible in that the

> state of the Attester, and the Appraisal Policy for Evidence,

Author: This text has been reverted.

might change immediately after the Evidence or Attestation Results were generated.
Freshness is a key component for enabling caching
and reuse of both Evidence and Attestation Results, which is especially
valuable in cases where their computation uses a substantial part of a
resource budget (e.g., an energy budget in constrained devices).
The goal is to narrow the time window to something the Verifier
(for Evidence) or Relying Party (for an Attestation Result) is willing to accept.

There are four common approaches to providing some assurance of freshness, effectively determining the epoch of the

> validity-span of, e.g., Evidence or Attestation Results.

Reviewer: "epoch of the validity-span" is hard to parse and understand.  Can this be rephrased in layman's terms?

Author: This text doesn't exist anymore.

entity (e.g., the Verifier for Evidence, or the Relying Party for an
Attestation Result).
This might be an appropriate choice in case the Attester does not have a reliable clock or time synchronisation is
otherwise impaired.
In this approach, a nonce is generated by a remote entity
(e.g., the Verifier for Evidence, or the Relying Party for an Attestation
Result), and the nonce is then signed and included along with the claims in
the Evidence or Attestation Result, so that the remote entity knows that the
claims were signed after the nonce was generated.

## Explicit Timekeeping using Synchronized Clocks

A second approach is to rely on synchronized clocks, and include a signed
timestamp along with the
claims in the Evidence or Attestation Result, so that the remote entity knows
that the claims were signed at that time, as long as it has some assurance that
the timestamp is correct.
This typically requires additional claims about
the signer's time synchronization mechanism in order to provide such assurance.
This allows associating a "rough" epoch to the Evidence or Attestation Result. 
In this case, the epoch is said to be "rough" because 

>     a) it is always bound to the set of Claims as a whole, as opposed to having Claim being separately annotated, and 

Reviewer: "having Claim" is bad grammar.
"having Claims"?  "having a Claim"? "having each Claim"?

But which approach is the one where each Claim can be separately annotated with a timestamp?   In the original text with only two approaches, it would be part of this one.

Author: This has been addressed using Ned's suggestion in c215569

    b) it is not possible to distinguish between time of collection from the time of creation of the Claims set.

## Interval Timekeeping using Nonces

A third approach is a variant of the first; the difference is that it allows
fine-grained epoch assignment.
In this approach an Attester's signature spans the nonce as well as an

> Attester's trusted timetick counter(s). 

Reviewer: For parity with previous subsection, this would then require adding text saying "This typically requires additional claims about the Attester's timetick counter mechanism in order to provide such assurance."

Author: This text doesn't exist anymore.

Signatures across subsequent Evidence which embeds incrementing values of
the trusted timetick counter(s) can be used to establish the time interval

> which has elasped since the nonce was received.

Reviewer: I don't yet follow why the time interval since nonce was received is interesting for anything.  Explain, or provide a forward reference if it's later.

Author: This text doesn't exist anymore.

> ## Distributed Interval Timekeeping using Local Clocks

Reviewer: I can't figure out how/why this is any different architecturally from nonces.  To me, the handle is just a nonce as far as I can tell from the below discussion.

Reviewer 2: It seems like that the handle is also a nonce, and the difference is that this nonce is from a third party (a centralized entity), not the Verifier or Relying Party.
I'm afraid this approach is more complex because you need to consider the time sync among three entities. For example, the time of receiving the handle among different entities may be different, therefore the handle expired time will be different. If the handle on the Verifier expires earlier than the Attester, the handle may already be expired while Verifier receives the Evidence from Attester.

Author: This text doesn't exist anymore.

A fourth approach places the onus of timekeeping on all involved parties,
but does not require the parties' clocks to be sychronized.
A centralized entity is required to distribute non-repeatable values, which are referred to as
Handles, to all involved parties in periodic intervals.

> Each Handle has a validity-span with an epoch that starts at the local time of

Reviewer: The term "validity-span" is not a normal term as far as I know

Author: This text doesn't exist anymore.

involved parties at which a distributed handle has been received.

> While only the source of the periodically sent handle must track the time interval, Attesters should also track is to avoid creating invalid Evidence when it can be avoided.

Reviewer: can't parse "track is to avoid"

Author: This text doesn't exist anymore.

> Distributed handles are used in place of a nonces for Evidence generation.

Reviewer 2: ```suggestion
Distributed handles are used in place of nonces for Evidence generation.
```

Author: This whole chunk of text is gone.

Verfiers have knowledge of the current valid
handle and keep track of its validity-span since the established epoch.
If a handle expired due to a time-out threshold, from the point of view
of Verifiers the time window of validity expired and the Evidence
cannot be considered fresh anymore.

## Claim Validity

In all approaches, it is important to note that the actual values in Claims
might have been generated long before the Claims are signed.
If so, it is the
signer's (the Attestation Environment's) responsibility to ensure that the values are still correct when they

> are signed.

Reviewer: I'm not sure I agree with this.  Shouldn't it be covered by the policy?  If an attester can only get a stale value and can't vouch for it not having been changed since, then putting that fact in the Evidence allows the verifier to make a decision appropriately.

There are two common approaches to providing some assurance of freshness.
The first approach is that a nonce is generated by a remote entity
(e.g., the Verifier for Evidence, or the Relying Party for an Attestation
Result), and the nonce is then signed and included along with the claims in
the Evidence or Attestation Result, so that the remote entity knows that the
claims were signed after the nonce was generated.

A second approach is to rely on synchronized clocks, and include a signed
timestamp (e.g., using {{-rats-tuda}}) along with the
claims in the Evidence or Attestation Result, so that the remote entity knows
that the claims were signed at that time, as long as it has some assurance that
the timestamp is correct.  This typically requires additional claims about
the signer's time synchronization mechanism in order to provide such assurance.

In either approach, it is important to note that the actual values in claims
might have been generated long before the claims are signed.  If so, it is the
signer's responsibility to ensure that the values are still correct when they
are signed.  For example, values might have been generated at boot, and then

> used in claims as long as the signer can guarantee that they cannot have changed

Reviewer 3: "could not have"

Author: Ned, you had another comment on this same sentence, which replaced it altogether.  Since your later suggestion has been already incorporated (c215569), I'm closing this.

> This is essential in deciding whether the included Claims can be considered fresh, meaning they still reflect the latest state of the Attester, and that any Attestation Result was generated using the latest Appraisal Policy for Evidence.

Reviewer 4: ```suggestion
This is essential in deciding whether the included Claims and their values can be considered fresh, meaning they still reflect the latest state of the Attester, and that any Attestation Result was generated using the latest Appraisal Policy for Evidence.
```

> Freshness is a key component  for enabling caching and reuse of both Evidence and Attestation Results, which is especially valuable in cases where their computation uses a substantial part of the energy budget (e.g., in constrained devices).

Reviewer 4: ```suggestion
Freshness is a key component  for enabling caching and reuse of both Evidence and Attestation Results, which is especially valuable in cases where their computation uses a substantial part of the resource budget (e.g., energy in constrained devices).
```

> This allows fine-grained epoch assignment, even on per-Claim basis if wanted, as well as the ability to distinguish time of collection from time of creation of Claims.

Reviewer 4: ```suggestion
This allows fine-grained epoch assignment, even on a per-Claim basis if wanted, as well as the ability to distinguish time of collection from time of creation of Claims.
```

> This allows fine-grained epoch assignment, even on per-Claim basis if wanted, as well as the ability to distinguish time of collection from time of creation of Claims.

Reviewer: ```suggestion
Timestamps can also be added on a per-Claim basis if wanted, to allow distinguishing the time of creation of Evidence from the time that a specific Claim value was generated.
```

> Timestamps can also be added on a per-Claim basis if wanted, to allow distinguishing the time of creation of Evidence from the time that a specific Claim value was generated.

Reviewer 3: "also", "if wanted", "of Evidence", "value" are not needed.
"allow distinguishing" change to "distinguish"

Author: fixed in c215569

> Clocks trustworthiness typically requires additional Claims about the signer's time synchronization mechanism.

Reviewer 3: "The clock's trustworthiness" - grammar.

Author: fixed in c215569

> A second approach places the onus of timekeeping solely on the remote entity (e.g., the Verifier for Evidence, or the Relying Party for an Attestation Result).

Reviewer 4: ```suggestion
A second approach places the onus of timekeeping solely on the appraising entity (e.g., the Verifier for Evidence, or the Relying Party for an Attestation Result).
```

> A second approach places the onus of timekeeping solely on the appraising entity (e.g., the Verifier for Evidence, or the Relying Party for an Attestation Result).

Reviewer 3: "the Verifier (for Evidence), or the Relying Party (for Attestation Results)."

Author: fixed in c215569

> In this approach, a non-predictable nonce is generated by the remote entity, and the nonce is then signed and included along with the Claims in the Evidence or Attestation Result.

Reviewer: ```suggestion
In this approach, a non-predictable nonce is sent by the remote entity, and the nonce is then signed and included along with the Claims in the Evidence or Attestation Result.
```

Author: Fixed in 34c9443

> In this approach, a non-predictable nonce is generated by the remote entity, and the nonce is then signed and included along with the Claims in the Evidence or Attestation Result.

Reviewer 4: ```suggestion
In this approach, a non-predictable nonce is generated by the appraising entity, and the nonce is then signed and included along with the Claims in the Evidence or Attestation Result.
```

> After checking the sent and received nonces are the same, the remote entity knows that the Claims were signed after the nonce was generated, effectively locking its clock to the produced payload.

Reviewer: ```suggestion
After the remote entity checks that the sent and received nonces are the same, the remote entity knows that the Claims were signed after the nonce was generated.
```

Author: Fixed in 34c9443

> After checking the sent and received nonces are the same, the remote entity knows that the Claims were signed after the nonce was generated, effectively locking its clock to the produced payload.

Reviewer 4: ```suggestion
After checking the sent and received nonces are the same, the appraising entity knows that the Claims were signed after the nonce was generated, effectively locking its clock to the produced payload.
```

> After checking the sent and received nonces are the same, the appraising entity knows that the Claims were signed after the nonce was generated, effectively locking its clock to the produced payload.

Reviewer: ```suggestion
After checking the sent and received nonces are the same, the appraising entity knows that the Claims were signed after the nonce was generated.
```

Author: Fixed in 35dc70a

> This allows associating a "rough" epoch to the Evidence or Attestation Result.

Reviewer 3: Replace "This" with "Timekeeping" ?

Author: Not sure: "this" is a bit more than just "timekeeping", it refers to "an approach that places the onus of timekeeping solely on the appraising entity".

Reviewer 3: Maybe substitute your text in place of 'this' since the pronoun reference seems ambiguous?

Author: That seems a bit redundant.  I'd like to achieve some kind of compression - obviously without introducing ambiguity.  Let me see if I can concoct something decent.

Author: Ned: please check the proposal in 20e3d7

> * it is always bound to the set of Claims as a whole, as opposed to having Claim being separately annotated, and

Reviewer 3: * The epoch applies to the entire claim set instead of a more granular association, and

Author: fixed in c215569

> This allows associating a "rough" epoch to the Evidence or Attestation Result.
> In this case the epoch is said to be rough because:
> 
> * it is always bound to the set of Claims as a whole, as opposed to having Claim being separately annotated, and
> * it is not possible to distinguish between time of collection from the time of creation of the Claims set.

Reviewer: ```suggestion
```

Author: I don't think we should ditch this.

The essential premise is on line 869. In particular, by using "solely" I wanted to discuss the case where the only clock that the appraiser can trust is its own.  The intent of line 870 is just to illustrate two common instances where the appraiser can't rely on the peer's clock.

The overall premise for this whole paragraph is that appraiser does _not_ consider any time assertion made the peer, and to determine what kind of information can be used in this case to evaluate freshness.

Clearly, once the peer's clock becomes usable - as in the example at line 880 - then finer-grained timestamping can become input to the evaluation of the freshness policy as well.

Hopefully I have made my intention clear and the ambiguities in the existing text can be fixed accordingly :-)

> * it is not possible to distinguish between time of collection from the time of creation of the Claims set.

Reviewer 3: * The time between the creation of Claims and the collection of Claims is indistinguishable.

Author: fixed in c215569

> The two basic mechanisms presented above can be extended and combined in more advanced schemes.

Reviewer 3: "Implicit and explicit timekeeping can be combined into hybrid mechanisms.

Author: fixed in c215569

> For example, if the clocks are trustworthy but not synchronized, first a nonce-based exchange may be used to determine the (rough) time offset between the involved peers, followed by any number of timestamp based exchanges.

Reviewer: ```suggestion
For example, if clocks exist and are considered trustworthy but are not synchronized, first a nonce-based exchange may be used to determine the (rough) time offset between the involved peers, followed by any number of timestamp based exchanges.
```

> In another scenario where a broadcast channel is shared by all the RATS entities (Attesters, Verifiers and Relying Parties), the nonce-based approach may be used to lock all parties to the same timeline without requiring synchronized clocks by having a central entity emit nonces at regular intervals and have the "current" nonce included in the produced Evidence or Attestation Result.

Reviewer 3: "Second, a broadcast channel shared by all Roles (Attester, Verifier, etc.), the nonce-based approach..."

Author: Not sure I understand this suggestion.  If the current text isn't good as-is, we may simplify it as "When a broadcast channel is shared by all Roles, the nonce-base approach [...]"

Reviewer 3: Trying to get the signposting to match up. In line 880 signposting uses "first" and in 881 signposting changes to "another". Normally, "second" follows "first" - it is just style and I don't feel strongly about it.

Author: Line 880 and 881 are two separate examples.

The example at line 880 is split in two phases: first [...], followed by [...].

I will remove "first" since it isn't necessary and I suspect it's probably bad English if it got you confused.

Author: Ned: please check the proposal in 20e3d7

> Whatever the approach, it is important to note that the actual values in Claims might have been generated long before the Claims are signed.

Reviewer 3: "It is important..."

Author: fixed in c215569

If so, it is the signer's responsibility to ensure that the values are still correct when they are signed.

> For example, values might have been generated at boot, and then used in Claims as long as the signer can guarantee that they cannot have changed since boot.

Reviewer 3: This is not an example imo. "For example, values generated at boot time might have been saved to secure storage until network connectivity is established to the remote Verifier and a nonce is obtained."

Author: fixed in c215569

Freshness is assessed based on a policy defined by the consuming entity,
Verifier or Relying Party, that compares the estimated epoch against an
"expiry" threshold defined locally to that policy.  There is, however, always a
race condition possible in that the state of the Attester, and the Appraisal
Policy for Evidence, might change immediately after the Evidence or Attestation
Result was generated.  The goal is merely to narrow their recentness to
something the Verifier (for Evidence) or Relying Party (for Attestation Result)
is willing to accept.  Freshness is a key component  for enabling caching and
reuse of both Evidence and Attestation Results, which is especially valuable in
cases where their computation uses a substantial part of the resource budget
(e.g., energy in constrained devices).

There are two common approaches for determining the epoch of an Evidence or
Attestation Result.

<!-- Explicit Timekeeping using Synchronized Clocks -->

The first approach is to rely on synchronized and trustworthy clocks, and
include a signed timestamp (see {{?I-D.birkholz-rats-tuda}}) along with the
Claims in the Evidence or Attestation Result.  Timestamps can be added on a

> per-Claim basis, to distinguish the time of creation from the time that a

Author: > Does 'the time of creation' mean the time of the creation of the Evidence or Attestation Result?

Yes.

> More explicit would be better.

OK

> per-Claim basis, to distinguish the time of creation from the time that a

Author: fixed in 8232e0a

specific Claim was generated.  The clock's trustworthiness typically requires
additional Claims about the signer's time synchronization mechanism.

<!-- Implicit Timekeeping using Nonces -->

A second approach places the onus of timekeeping solely on the appraising
entity, e.g., the Verifier (for Evidence), or the Relying Party (for

> Attestation Results), and might be suitable in case the Attester does not have

Reviewer 2: I think 'the Attester' here can't cover the case where the Relying Party sends the nonce to the Verifier for retrieving the Attestation Result. Maybe change 'the Attester' to 'the producing entity'?

Author: 1. You are right in that it is not limited to the Attester.  I wanted to use the it as the typical example because it's more likely that a constrained Attester doesn't have a very reliable clock.  What about: `and might be suitable, for example, in case the Attester does not have [...]`?

2. The current paragraph looks wrong to me.  When it says:
```
[...] solely on the appraising entity, e.g., the Verifier (for Evidence),
or the Relying Party (for Attestation Results),`
```
should instead say:
```
[...] solely on the remote attestation protocol initiator, e.g., the
Relying Party (for Evidence or Attestation Results) and the Attester
(for Attestation Results).
```

Policy for Evidence, might change immediately after the Evidence or Attestation
Result was generated.  The goal is merely to narrow their recentness to
something the Verifier (for Evidence) or Relying Party (for Attestation Result)
is willing to accept.  Freshness is a key component for enabling caching and
reuse of both Evidence and Attestation Results, which is especially valuable in
cases where their computation uses a substantial part of the resource budget
(e.g., energy in constrained devices).

There are two common approaches for determining the epoch of an Evidence or
Attestation Result.

<!-- Explicit Timekeeping using Synchronized Clocks -->

The first approach is to rely on synchronized and trustworthy clocks, and
include a signed timestamp (see {{?I-D.birkholz-rats-tuda}}) along with the
Claims in the Evidence or Attestation Result.  Timestamps can be added on a
per-Claim basis, to distinguish the time of creation of Evidence or Attestation
Result from the time that a specific Claim was generated.  The clock's
trustworthiness typically requires additional Claims about the signer's time
synchronization mechanism.

<!-- Implicit Timekeeping using Nonces -->

> A second approach places the onus of timekeeping solely on the appraising
> entity, e.g., the Verifier (for Evidence), or the Relying Party (for
> Attestation Results), and might be suitable in case the Attester does not have

Reviewer 4: ```suggestion
A second approach places the onus of timekeeping solely on the appraising
entity, e.g., the Verifier (for Evidence), or the Relying Party (for
Attestation Results), and might be suitable, for example, in case the Attester does not have
```

The table below defines a number of relevant events, with an ID that
is used in subsequent diagrams.  The times of said events might be
defined in terms of an absolute clock time such as Coordinated Universal Time, 
or might be defined relative to some other timestamp or timeticks counter with different timescales.

| ID | Event                       | Explanation of event
|----|-----------------------------|-----------------------
| VG | Value generation            | A value to appear in a Claim was created.
| AA | Attester awareness          | An Attesting Environment starts to be aware of a new/changed Claim value.
| CC | Claim Collection            | An Attesting Environment collects a new/changed Claim value to appear in Evidence.

> | HD | Handle distribution         | A centrally generated identifier for time-bound recentness across a domain of devices is successfully received by Attesters (and Verifiers).

Reviewer 2: I think these new-added events had been questioned before, that the examples below don't use them.

Author: This chunk of text is gone.
